File: src/data/usageReport.ts
import { orderedProviderKeys, providerLabel } from '../display/providers.js';
import {
  computePreviousUsageWindow,
  computeUsageWindow,
  type UsagePeriod,
  type UsageWindow,
} from '../display/usagePeriods.js';
import { aggregateModelsByDayMap } from './aggregate.js';
import { isUsageNotConfigured, usageEmptyMessage, usageEmptyWindowMessage } from './emptyState.js';
import { compareByCostThenTokens } from './sort.js';
import type { ProviderData } from './types.js';
import { loadMergedProviderData } from './usageData.js';

export interface UsageReportOptions {
  period: UsagePeriod;
  providers?: string[];
  from?: string;
  to?: string;
  n?: number;
}

export interface UsageReportRow {
  model: string;
  inputTokens: number;
  outputTokens: number;
  tokens: number;
  costUSD: number;
  hasCost: boolean;
}

export interface UsageReportProvider {
  key: string;
  label: string;
  rows: UsageReportRow[];
  subtotalTokens: number;
  subtotalCostUSD: number;
  subtotalHasCost: boolean;
}

export interface UsageReportTotals {
  inputTokens: number;
  outputTokens: number;
  tokens: number;
  costUSD: number;
  hasCost: boolean;
}

export interface UsageReport {
  windowLabel: string;
  providers: UsageReportProvider[];
  totals: UsageReportTotals;
  rowCount: number;
}

export interface UsageComparisonMetric {
  current: number;
  previous: number;
  delta: number;
  percentChange: number | null;
}

export interface UsageModelComparison {
  providerKey: string;
  providerLabel: string;
  model: string;
  tokens: UsageComparisonMetric;
  costUSD: UsageComparisonMetric;
  hasCost: boolean;
}

export interface UsageComparison {
  previousWindowLabel: string;
  totals: {
    tokens: UsageComparisonMetric;
    costUSD: UsageComparisonMetric;
    hasCost: boolean;
  };
  models: UsageModelComparison[];
}

export interface UsageComparisonReport {
  current: UsageReport;
  previous: UsageReport;
  comparison: UsageComparison;
}

function buildUsageReportFromData(providerData: ProviderData, window: UsageWindow): UsageReport {
  const ordered = orderedProviderKeys(providerData);
  const providers: UsageReportProvider[] = [];
  const totals: UsageReportTotals = {
    inputTokens: 0,
    outputTokens: 0,
    tokens: 0,
    costUSD: 0,
    hasCost: false,
  };
  let rowCount = 0;

  for (const key of ordered) {
    const dayMap = providerData[key];
    if (!dayMap) continue;
    const byModel = aggregateModelsByDayMap(dayMap, { start: window.start, end: window.end });

    const rows: UsageReportRow[] = [];
    let subtotalTokens = 0;
    let subtotalCostUSD = 0;
    let isSubtotalHasCost = false;

    for (const [model, agg] of byModel) {
      const tokens = agg.inputTokens + agg.outputTokens;
      if (tokens === 0 && !agg.hasCost) continue;
      rows.push({
        model,
        inputTokens: agg.inputTokens,
        outputTokens: agg.outputTokens,
        tokens,
        costUSD: agg.hasCost ? agg.costUSD : 0,
        hasCost: agg.hasCost,
      });
      subtotalTokens += tokens;
      totals.inputTokens += agg.inputTokens;
      totals.outputTokens += agg.outputTokens;
      if (agg.hasCost) {
        subtotalCostUSD += agg.costUSD;
        isSubtotalHasCost = true;
        totals.costUSD += agg.costUSD;
        totals.hasCost = true;
      }
    }

    if (rows.length === 0) continue;
    rows.sort((a, b) => compareByCostThenTokens(a, b));
    rowCount += rows.length;
    providers.push({
      key,
      label: providerLabel(key),
      rows,
      subtotalTokens,
      subtotalCostUSD,
      subtotalHasCost: isSubtotalHasCost,
    });
  }

  totals.tokens = totals.inputTokens + totals.outputTokens;
  return { windowLabel: window.label, providers, totals, rowCount };
}

function comparisonMetric(current: number, previous: number): UsageComparisonMetric {
  return {
    current,
    previous,
    delta: current - previous,
    percentChange: previous === 0 ? null : ((current - previous) / previous) * 100,
  };
}

function rowsByProviderAndModel(report: UsageReport): Map<string, UsageReportRow> {
  const rows = new Map<string, UsageReportRow>();
  for (const provider of report.providers) {
    for (const row of provider.rows) {
      rows.set(`${provider.key}\0${row.model}`, row);
    }
  }
  return rows;
}

function compareUsageReports(current: UsageReport, previous: UsageReport): UsageComparison {
  const currentRows = rowsByProviderAndModel(current);
  const previousRows = rowsByProviderAndModel(previous);
  const keys = new Set([...currentRows.keys(), ...previousRows.keys()]);
  const models: UsageModelComparison[] = [];

  for (const key of keys) {
    const separator = key.indexOf('\0');
    const providerKey = key.slice(0, separator);
    const model = key.slice(separator + 1);
    const currentRow = currentRows.get(key);
    const previousRow = previousRows.get(key);
    models.push({
      providerKey,
      providerLabel: providerLabel(providerKey),
      model,
      tokens: comparisonMetric(currentRow?.tokens ?? 0, previousRow?.tokens ?? 0),
      costUSD: comparisonMetric(currentRow?.costUSD ?? 0, previousRow?.costUSD ?? 0),
      hasCost: (currentRow?.hasCost ?? false) || (previousRow?.hasCost ?? false),
    });
  }

  models.sort((a, b) => {
    const costDifference = Math.abs(b.costUSD.delta) - Math.abs(a.costUSD.delta);
    if (costDifference !== 0) return costDifference;
    const tokenDifference = Math.abs(b.tokens.delta) - Math.abs(a.tokens.delta);
    if (tokenDifference !== 0) return tokenDifference;
    return `${a.providerKey}\0${a.model}`.localeCompare(`${b.providerKey}\0${b.model}`);
  });

  return {
    previousWindowLabel: previous.windowLabel,
    totals: {
      tokens: comparisonMetric(current.totals.tokens, previous.totals.tokens),
      costUSD: comparisonMetric(current.totals.costUSD, previous.totals.costUSD),
      hasCost: current.totals.hasCost || previous.totals.hasCost,
    },
    models,
  };
}

/**
 * Load merged provider data and build a structured per-provider / per-model
 * usage report for a time window. Returns null when no provider data is
 * available at all (caller should show the "not configured" hint); a report
 * with rowCount === 0 means data exists but nothing falls in the window.
 *
 * Shared by the `usage` (terminal table) and `export` (PDF receipt) commands so
 * both render from a single source of truth.
 */
export async function buildUsageReport(options: UsageReportOptions): Promise<UsageReport | null> {
  const loaded = await loadMergedProviderData({ providers: options.providers });
  if (!loaded) return null;

  const window = computeUsageWindow(options);
  return buildUsageReportFromData(loaded.providerData, window);
}

export async function buildUsageComparison(
  options: UsageReportOptions,
): Promise<UsageComparisonReport | null> {
  const currentWindow = computeUsageWindow(options);
  const previousWindow = computePreviousUsageWindow(options, currentWindow);
  const loaded = await loadMergedProviderData({ providers: options.providers });
  if (!loaded) return null;

  const current = buildUsageReportFromData(loaded.providerData, currentWindow);
  const previous = buildUsageReportFromData(loaded.providerData, previousWindow);
  return { current, previous, comparison: compareUsageReports(current, previous) };
}

/**
 * Message to print when a report has nothing to render, or null when it does.
 * Distinguishes "no data at all / not configured" (report is null) from "data
 * exists but the window is empty" (rowCount === 0). Shared by the `usage` and
 * `export` commands so both report empty states identically.
 */
export function emptyReportMessage(report: UsageReport | null): string | null {
  if (!report) return usageEmptyMessage(isUsageNotConfigured());
  if (report.rowCount === 0) return usageEmptyWindowMessage(report.windowLabel);
  return null;
}

File: crates/aitrack-core/src/data/usage_report.rs
//! Usage report generation and comparison.

use super::aggregate::{AggregateModelsFilter, aggregate_models_by_day_map};
use super::types::ProviderData;
use super::usage_data::{LoadOptions, load_merged_provider_data};
use super::usage_periods::{
    UsagePeriod, UsageWindow, UsageWindowOptions, compute_previous_usage_window,
    compute_usage_window,
};

#[derive(Debug, Clone)]
pub struct UsageReportOptions {
    pub period: UsagePeriod,
    pub providers: Option<Vec<String>>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub n: Option<usize>,
    pub refresh_live: bool,
}

#[derive(Debug, Clone)]
pub struct UsageReportRow {
    pub model: String,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub tokens: u64,
    pub cached_input_tokens: u64,
    pub has_cached: bool,
    pub cost_usd: f64,
    pub has_cost: bool,
}

#[derive(Debug, Clone)]
pub struct UsageReportProvider {
    pub key: String,
    pub label: String,
    pub rows: Vec<UsageReportRow>,
    pub subtotal_tokens: u64,
    pub subtotal_cost_usd: f64,
    pub subtotal_has_cost: bool,
}

#[derive(Debug, Clone)]
pub struct UsageReportTotals {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub tokens: u64,
    pub cached_input_tokens: u64,
    pub has_cached: bool,
    pub cost_usd: f64,
    pub has_cost: bool,
}

#[derive(Debug, Clone)]
pub struct UsageReport {
    pub window_label: String,
    pub providers: Vec<UsageReportProvider>,
    pub totals: UsageReportTotals,
    pub row_count: usize,
}

#[derive(Debug, Clone)]
pub struct UsageComparisonMetric {
    pub current: f64,
    pub previous: f64,
    pub delta: f64,
    pub percent_change: Option<f64>,
}

#[derive(Debug, Clone)]
pub struct UsageModelComparison {
    pub provider_key: String,
    pub provider_label: String,
    pub model: String,
    pub tokens: UsageComparisonMetric,
    pub cost_usd: UsageComparisonMetric,
    pub has_cost: bool,
}

#[derive(Debug, Clone)]
pub struct UsageComparison {
    pub previous_window_label: String,
    pub totals: UsageComparisonTotals,
    pub models: Vec<UsageModelComparison>,
}

#[derive(Debug, Clone)]
pub struct UsageComparisonTotals {
    pub tokens: UsageComparisonMetric,
    pub cost_usd: UsageComparisonMetric,
    pub has_cost: bool,
}

#[derive(Debug, Clone)]
pub struct UsageComparisonReport {
    pub current: UsageReport,
    pub previous: UsageReport,
    pub comparison: UsageComparison,
}

fn build_usage_report_from_data(provider_data: &ProviderData, window: &UsageWindow) -> UsageReport {
    let filter = AggregateModelsFilter {
        start: Some(window.start.clone()),
        end: Some(window.end.clone()),
        year: None,
    };

    let mut providers = Vec::new();
    let mut totals = UsageReportTotals {
        input_tokens: 0,
        output_tokens: 0,
        tokens: 0,
        cached_input_tokens: 0,
        has_cached: false,
        cost_usd: 0.0,
        has_cost: false,
    };
    let mut row_count = 0;

    for (provider_key, day_map) in provider_data {
        let models = aggregate_models_by_day_map(day_map, Some(&filter));
        if models.is_empty() {
            continue;
        }

        let mut rows = Vec::new();
        let mut subtotal_tokens = 0;
        let mut subtotal_cost_usd = 0.0;
        let mut subtotal_has_cost = false;

        for (model, agg) in models {
            let tokens = agg.input_tokens + agg.output_tokens;
            rows.push(UsageReportRow {
                model,
                input_tokens: agg.input_tokens,
                output_tokens: agg.output_tokens,
                tokens,
                cached_input_tokens: agg.cached_input_tokens,
                has_cached: agg.has_cached,
                cost_usd: agg.cost_usd,
                has_cost: agg.has_cost,
            });

            subtotal_tokens += tokens;
            if agg.has_cost {
                subtotal_cost_usd += agg.cost_usd;
                subtotal_has_cost = true;
            }
        }

        row_count += rows.len();

        providers.push(UsageReportProvider {
            key: provider_key.clone(),
            label: provider_key.clone(), // Will use provider registry later
            rows,
            subtotal_tokens,
            subtotal_cost_usd,
            subtotal_has_cost,
        });

        // Update totals
        totals.tokens += subtotal_tokens;
        if subtotal_has_cost {
            totals.cost_usd += subtotal_cost_usd;
            totals.has_cost = true;
        }
    }

    UsageReport {
        window_label: window.label.clone(),
        providers,
        totals,
        row_count,
    }
}

pub fn build_usage_report(options: &UsageReportOptions) -> anyhow::Result<UsageReport> {
    let window_options = UsageWindowOptions {
        period: options.period,
        from: options.from.clone(),
        to: options.to.clone(),
        n: options.n,
    };

    let window = compute_usage_window(&window_options);

    let load_opts = LoadOptions {
        providers: options.providers.clone(),
        refresh_live: options.refresh_live,
    };

    let loaded = load_merged_provider_data(&load_opts)?;

    Ok(build_usage_report_from_data(&loaded.provider_data, &window))
}

pub fn build_usage_comparison(
    options: &UsageReportOptions,
) -> anyhow::Result<UsageComparisonReport> {
    let window_options = UsageWindowOptions {
        period: options.period,
        from: options.from.clone(),
        to: options.to.clone(),
        n: options.n,
    };

    let current_window = compute_usage_window(&window_options);
    let previous_window = compute_previous_usage_window(&window_options, Some(&current_window));

    let load_opts = LoadOptions {
        providers: options.providers.clone(),
        refresh_live: options.refresh_live,
    };

    let loaded = load_merged_provider_data(&load_opts)?;

    let current = build_usage_report_from_data(&loaded.provider_data, &current_window);
    let previous = build_usage_report_from_data(&loaded.provider_data, &previous_window);

    // Build comparison (simplified for now)
    let tokens_metric = UsageComparisonMetric {
        current: current.totals.tokens as f64,
        previous: previous.totals.tokens as f64,
        delta: current.totals.tokens as f64 - previous.totals.tokens as f64,
        percent_change: if previous.totals.tokens > 0 {
            Some(
                ((current.totals.tokens as f64 - previous.totals.tokens as f64)
                    / previous.totals.tokens as f64)
                    * 100.0,
            )
        } else {
            None
        },
    };

    let cost_metric = UsageComparisonMetric {
        current: current.totals.cost_usd,
        previous: previous.totals.cost_usd,
        delta: current.totals.cost_usd - previous.totals.cost_usd,
        percent_change: if previous.totals.cost_usd > 0.0 {
            Some(
                ((current.totals.cost_usd - previous.totals.cost_usd) / previous.totals.cost_usd)
                    * 100.0,
            )
        } else {
            None
        },
    };

    let comparison = UsageComparison {
        previous_window_label: previous_window.label,
        totals: UsageComparisonTotals {
            tokens: tokens_metric,
            cost_usd: cost_metric,
            has_cost: current.totals.has_cost,
        },
        models: Vec::new(), // Simplified for now
    };

    Ok(UsageComparisonReport {
        current,
        previous,
        comparison,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_usage_report_struct() {
        let report = UsageReport {
            window_label: "test".to_string(),
            providers: Vec::new(),
            totals: UsageReportTotals {
                input_tokens: 0,
                output_tokens: 0,
                tokens: 0,
                cached_input_tokens: 0,
                has_cached: false,
                cost_usd: 0.0,
                has_cost: false,
            },
            row_count: 0,
        };
        assert_eq!(report.window_label, "test");
    }
}
